### ISSUE TITLE
Backbone producing nan

### ISSUE THREAD

**Issue by i-pan** [open-mmlab/mmdetection#570]
During training on a custom dataset, I noticed that `loss_reg` was nan. After looking into it further, I realized that when calling the backbone to extract features was resulting in tensors filled with nan for the last 3 layers. This is occurring across all available backbones and detectors.

The problem seems to be here (taken from mmdet.models.backbones.ResNet): 
```
    def forward(self, x):
        x = self.conv1(x)
        x = self.norm1(x)
        x = self.relu(x)
        x = self.maxpool(x)
        outs = []
        for i, layer_name in enumerate(self.res_layers):
            res_layer = getattr(self, layer_name)
            x = res_layer(x)
            if i in self.out_indices:
                outs.append(x)
        return tuple(outs)
```
Anyone encounter a similar problem? It's interesting because the first `res_layer` produces a output but subsequent layers produce all nans.

**Comment by hellock** [comment]
NaN is usually caused by the improper gradient, which will harm the network weights.

**Comment by soon-yau** [comment]
> NaN is usually caused by the improper gradient, which will harm the network weights.

I'm seeing similar problem. The FasterRCNN-ResnetX was training alright in the first few hundreds steps, then once NaN starts to appear, it never recover. Any tips on things to try to get rid of NaN?

``
2019-06-19 08:17:31,414 - INFO - Epoch [1][75000/217878]        lr: 0.00300, eta: 29 days, 9:59:30, time: 0.589, data_time: 0.005, memory: 8595, loss_rpn_cls: 0.1343, loss_rpn_bbox: 0.0180, loss_cls
: nan, acc: 27.5210, loss_bbox: nan, loss: nan
Warning: NaN or Inf found in input tensor.
Warning: NaN or Inf found in input tensor.
Warning: NaN or Inf found in input tensor.
``

**Comment by SnowRipple** [comment]
I am seeing Nans with retinanext as well

**Comment by soon-yau** [comment]
I did some more debugging by adding autograd.detect_anomaly() in mmdet/core/utils/dist_utils.py
``
from torch import autograd
...
    def after_train_iter(self, runner):
        with autograd.detect_anomaly():
            runner.optimizer.zero_grad()
            runner.outputs['loss'].backward()
            allreduce_grads(runner.model, self.coalesce, self.bucket_size_mb)
            if self.grad_clip is not None:
                self.clip_grads(runner.model.parameters())
            runner.optimizer.step()

``
Finally, I traced my problem to be that the bbox 'target' in smooth_l1_loss() contains NaN. As the bboxes annotation look alright on my data, so I'm not sure how the NaN got there after all the transforms. I worked around the problem by checking for NaN and setting them to 0.

**Comment by liuliu66** [comment]
@soon-yau Hi, I meet the similar problem. could you please tell me how to check the grad and how to set them to 0?

**Comment by analyticsbot** [comment]
One way I found this could be solved is to add the grad_norm to the train config. However, when running distributed training, this value needs to be higher. Otherwise the training runs into nan

**Comment by bc-arl** [comment]
Hi, I am also having this problem. @analyticsbot how much higher did you set the grad_norm to be? I am doing 2 GPUs with batch size 20, learning rate 8e-5, 2000 warmup iters, and grad_norm 35 (type 2) and I run into NaNs after about 50 epochs. After the NaNs start the training does not recover. Worth noting my dataset is quite large (50k images) and I checked my annotations, they are correct.
